ssl/ssl_lib.c

```diff
@@ -2205,20 +2212,41 @@ long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_SESS_CACHE_SIZE:
         if (larg < 0)
             return 0;
-        l = (long)ctx->session_cache_size;
-        ctx->session_cache_size = (size_t)larg;
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = (long)ctx->session_cache->size;
+        ctx->session_cache->size = (size_t)larg;
+        CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+        CRYPTO_THREAD_unlock(ctx->lock);
         return l;
     case SSL_CTRL_GET_SESS_CACHE_SIZE:
-        return (long)(ctx->session_cache_size);
+        CRYPTO_THREAD_read_lock(ctx->lock);
+        l = (long)ctx->session_cache->size;
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return l;
+
     case SSL_CTRL_SET_SESS_CACHE_MODE:
-        l = ctx->session_cache_mode;
-        ctx->session_cache_mode = larg;
-        return (l);
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = ctx->session_cache->mode;
+        ctx->session_cache->mode = larg;
```

Comment:
I am not sure if different sessions with different mode can share the same session.
Maybe the mode should not be changeable after a cache is assigned to a session.
Then a lot of the locks here could be avoided.
Or maybe the mode is a true property of the ctx, not sure...